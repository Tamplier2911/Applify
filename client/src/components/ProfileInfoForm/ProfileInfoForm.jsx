// import "./ProfileInfoForm.scss";
import React, { useState, useEffect } from "react";

// validator
import emailValidator from "../../utils/EmailValidator/emailValidator";

// redux
import { connect } from "react-redux";
import { createStructuredSelector } from "reselect";
import { selectCurrentLanguage } from "../../redux/lang/lang.selectors";
import { openModal } from "../../redux/modal/modal.actions";
import { updateUserDataStart } from "../../redux/auth/auth.actions";

// components
import FormInput from "../FormInput/FormInput";
import FileInput from "../FileInput/FileInput";
import Button from "../Button/Button";

// JS Rendering CSS
import {
  ProfileInfoFormContainer,
  ProfileInfoFormTitle,
  ProfileInfoFormElement
} from "./ProfileInfoFormStyles";

// component constants
import profileInfoFormData from "../../utils/ComponentProfileInfoFormConstants/componentProfileInfoFormConstants";

const ProfileInfoForm = ({
  name,
  email,
  photo,
  lang,
  openModal,
  updateUserDataStart
}) => {
  const {
    profileInfoFormTitle,
    profileInfoFormName,
    profileInfoFormEmail,
    profileInfoFormFileUpload,
    profileInfoFormFileButton
  } = profileInfoFormData[lang];

  const [profileInfo, setProfileInfo] = useState({
    profileName: "",
    profileEmail: "",
    profileImage: null,
    profileOldImage: photo
  });

  const { profileName, profileEmail, profileImage } = profileInfo;

  useEffect(() => {
    setProfileInfo({
      profileName: name,
      profileEmail: email,
      profileImage: null,
      profileOldImage: photo
    });
  }, [email, name, photo]);

  const onInputChange = e => {
    const { name, value } = e.target;
    if (name === "profileImage") {
      const [file] = e.target.files;
      setProfileInfo({ ...profileInfo, [name]: file });
    } else {
      setProfileInfo({ ...profileInfo, [name]: value });
    }
  };

  const onFormSubmit = e => {
    e.preventDefault();
    // validate data
    // validate type of image to be type: "image/..."
    if (!emailValidator(profileEmail) || !profileName)
      return openModal({
        header: "Attention!",
        content: "In order to update user data, please, enter name and email."
      });

    if (profileImage) {
      if (
        !profileImage.type.startsWith("image") ||
        parseInt(profileImage.size / 1024 / 1024) > 5
      ) {
        return openModal({
          header: "Attention!",
          content:
            "File must be type of 'image'. Note: max allowed file size - 5mb."
        });
      }
    }
    updateUserDataStart(profileInfo);
  };

  return (
    <ProfileInfoFormContainer>
      <ProfileInfoFormTitle lang={lang}>
        {profileInfoFormTitle}
      </ProfileInfoFormTitle>
      <ProfileInfoFormElement
        autoComplete="off"
        onSubmit={e => onFormSubmit(e)}
      >
        <FormInput
          onInputChange={e => onInputChange(e)}
          value={profileName}
          label={profileInfoFormName}
          name="profileName"
          type="text"
          required
        />
        <FormInput
          onInputChange={e => onInputChange(e)}
          value={profileEmail}
          label={profileInfoFormEmail}
          name="profileEmail"
          type="email"
          required
        />
        <FileInput
          onInputChange={e => onInputChange(e)}
          label={profileInfoFormFileUpload}
          image={profileImage ? 1 : 0}
          name="profileImage"
          type="file"
          id="profileInfoFormFileInput"
        />
        <Button type="submit" value={profileInfoFormFileButton} />
      </ProfileInfoFormElement>
    </ProfileInfoFormContainer>
  );
};

const mapStateToProps = createStructuredSelector({
  lang: selectCurrentLanguage
});

export default connect(mapStateToProps, { openModal, updateUserDataStart })(
  ProfileInfoForm
);
