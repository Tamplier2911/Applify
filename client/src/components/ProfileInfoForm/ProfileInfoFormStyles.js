import styled from "styled-components";

const getFont = props => {
  const { lang } = props;
  if (lang === "eng") {
    return `font-family: var(--font-title);`;
  } else if (lang === "rus" || lang === "ukr") {
    return `font-family: var(--font-cyrillic);`;
  }
  return `font-family: var(--font-title);`;
};

export const ProfileInfoFormContainer = styled.div`
  display: grid;
  grid-row-gap: 4rem;
  grid-auto-rows: min-content;

  padding: 2rem;
  box-shadow: 0 0.1rem 0.5rem var(--cl-font);
`;

export const ProfileInfoFormTitle = styled.h3`
  ${getFont}
  font-weight: 400;
  font-size: 1.7rem;

  margin-top: 2rem;
  @media only screen and (max-width: 425px) {
    //   margin-top: 5rem;
  }
`;

export const ProfileInfoFormElement = styled.form`
  display: grid;
  grid-row-gap: 2rem;
`;
