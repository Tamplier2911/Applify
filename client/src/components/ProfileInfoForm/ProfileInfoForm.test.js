import React from "react";
import toJson from "enzyme-to-json";

// component
import ProfileInfoForm from "./ProfileInfoForm";

// utils
import {
  setShallowC,
  findByTestAttr,
  storeFactory,
  checkProps,
} from "../../tests/testUtils";

describe("<ProfileInfoForm />", () => {
  const initialStore = {
    lang: {
      languagePanelHidden: true,
      currentLanguage: "eng",
    },
  };
  const store = storeFactory(initialStore);

  let wrapper;
  const defProps = {
    name: "Thomas",
    email: "ex@maple.com",
    photo: "profile.jpg",
  };
  beforeEach(() => {
    wrapper = setShallowC(<ProfileInfoForm store={store} {...defProps} />);
  });

  it("renders without an error", () => {
    const profileInfoForm = findByTestAttr(wrapper, "profile-info-form");
    expect(profileInfoForm.length).toBe(1);
  });

  it("renders - children profile info form element", () => {
    const profileInfoFormElement = findByTestAttr(
      wrapper,
      "profile-info-form-element"
    );
    expect(profileInfoFormElement.length).toBe(1);
  });

  it("recieves expected props", () => {
    const expectedProps = {
      name: "Thomas",
      email: "ex@maple.com",
      photo: "profile.jpg",
    };
    const err = checkProps(ProfileInfoForm, expectedProps, "props");
    expect(err).toBeUndefined();
  });

  it("matches snapshot", () => {
    const profileInfoForm = findByTestAttr(wrapper, "profile-info-form");
    expect(toJson(profileInfoForm)).toMatchSnapshot();
  });
});
