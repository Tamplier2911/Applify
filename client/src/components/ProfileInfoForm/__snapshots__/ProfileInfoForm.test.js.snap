// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<ProfileInfoForm /> matches snapshot 1`] = `
<Styled(ScrollAnimation)
  animateIn="bounceInRight"
  animateOnce={true}
  data-test="profile-info-form"
>
  <styled.h3
    lang="eng"
  >
    Change Credentials:
  </styled.h3>
  <styled.form
    autoComplete="off"
    data-test="profile-info-form-element"
    onSubmit={[Function]}
  >
    <FormInput
      label="Name"
      name="profileName"
      onInputChange={[Function]}
      required={true}
      type="text"
      value=""
    />
    <FormInput
      label="Email"
      name="profileEmail"
      onInputChange={[Function]}
      required={true}
      type="email"
      value=""
    />
    <FileInput
      id="profileInfoFormFileInput"
      image={0}
      label="Upload Image"
      name="profileImage"
      onInputChange={[Function]}
      type="file"
    />
    <Button
      type="submit"
      value="Save Changes"
    />
  </styled.form>
</Styled(ScrollAnimation)>
`;
